syntax = "proto3";
package org.roylance.yadel;

enum UIYadelRequestType {
    REPORT_DAGS = 0;
    DELETE_DAG = 1;
    GET_DAG_STATUS = 2;
    REPORT_DAGS_ACTIVE = 3;
}

message UIYadelRequest {
    UIYadelRequestType request_type = 1;
    string dag_id = 2;
    string token = 3;
    string user_name = 4;
    string password = 5;
}

message UIYadelResponse {
    string content = 1;
    UIDagReport report = 2;
    UIDag dag = 3;
    repeated UIWorkerConfiguration configurations = 4;
}

message UIDagReport {
    repeated UIWorkerConfiguration workers = 1;
    repeated UIDag dags = 2;
    string used_manager_memory = 3;
    string total_manager_memory = 4;
    int32 active_dags = 5;
    int32 unprocessed_dags = 6;
    int32 saved_dags = 7;
}

message UIDag {
    string id = 1;
    string display = 2;
    repeated UINode nodes = 3;
    repeated UIEdge edges = 4;
    bool is_completed = 5;
    bool is_processing = 6;
    bool is_error = 7;
    repeated string logs = 8;
    int32 number_completed = 9;
    int32 number_processing = 10;
    int32 number_errored = 11;
    int32 number_unprocessed = 12;
    repeated UIDag children = 13;
}

message UIEdge {
    string node_id_1 = 1;
    string node_id_2 = 2;
}

message UINode {
    string id = 1;
    string display = 2;
    int64 execution_date = 4;
    int64 start_date = 5;
    int64 end_date = 6;
    int64 duration = 7;
    bool is_completed = 8;
    bool is_processing = 9;
    bool is_error = 10;
    repeated UILog logs = 11;
    bool is_waiting_for_another_dag_task = 12;
}

message UILog {
    string id = 1;
    string message = 2;
}

message UIWorkerConfiguration {
    string ip = 1;
    string port = 2;
    string host = 3;
    string initialized_time = 4;
    UIWorkerState state = 5;
    string task_display = 6;
    string dag_display = 7;
    uint64 minutes_before_task_reset = 8;
    string task_start_time = 9;
    string task_working_time_display = 10;
}

enum UIWorkerState {
    CURRENTLY_WORKING = 0;
    CURRENTLY_IDLE = 1;
}